### ISSUE TITLE
[Veldrid.ImGui] Using custom fonts is currently not possible

### ISSUE THREAD

**Issue by SomeAnon42** [mellinoe/veldrid#154]
``ImGuiRenderer`` calls ``ImGui.NewFrame`` in the constructor, which locks the font atlas and makes it impossible to load and use a custom font.
ImGui has a debug assertion that's meant to prevent you from attempting that:
```cpp
// imgui_draw.cpp, line 1600
IM_ASSERT(!Locked && "Cannot modify a locked ImFontAtlas between NewFrame() and EndFrame/Render()!");
```

It's likely not there in the release build, so nothing bad seemingly happens in that situation when using ImGui.NET. However, the font texture doesn't actually get updated (not even if you call ``RecreateFontDeviceTexture`` afterwards). It's simply not supposed to be touched mid-frame.

Something needs to be done in order to allow the user to perform additional setup before ``NewFrame`` is called.

**Comment by mellinoe** [comment]
Indeed, this is a bit of an oversight, and it doesn't seem like there is an easy way to work around this restriction. I'd like to fix this, but I don't have a ton of time to brainstorm the best approach. The options that come to mind:

1. Add an option to the ImGuiRenderer constructor that prevents `NewFrame` from being called initially. It would be up to you to call that at some point afterwards, but before you start drawing stuff with ImGui. Seems pretty straightforward. Something like `ImGuiRendererOptions.ManualNewFrame`.
2. Add an ImGuiRenderer constructor that takes a list of fonts to load, and then add an accessor to retrieve the fonts that have been loaded. I don't like this as much because I'd prefer to let folks directly use the ImGui functions, since fonts can be loaded from a variety of sources and with a variety of options.
3. Add an ImGuiRenderer constructor that has some kind of callback that lets you do custom init before `NewFrame` is called. Seems unnecessary, but you don't have to worry about calling ImGui.NewFrame yourself.

@SomeAnon42  Thoughts?

**Comment by SomeAnon42** [comment]
I really don't think there's a need for callbacks or anything like that. Option 1 seems just fine, though I'm not sure adding a new enum or struct for this is even necessary when a simple bool flag can do the job. Well, not that it matters that much to me.

**Comment by mellinoe** [comment]
Actually, there is a fairly simple workaround for the time being, but it relies on the change I just pushed to master: https://github.com/mellinoe/veldrid/commit/db2eab0bdcf64f03de664bb7253f01ad81d4b7ee. ImGuiRenderer.RecreateFontTexture was recreating the Texture object, but not creating a new ResourceSet, so it wasn't actually being used in subsequent renders. With that change, you can do something like this to load different fonts:

```C#
ImGui.EndFrame();
var io = ImGui.GetIO();
io.Fonts.Clear();
io.Fonts.AddFontFromFileTTF("C:/Windows/Fonts/ArialUni.ttf", 14);
// Load other fonts
renderer.RecreateFontDeviceTexture(_gd);
ImGui.NewFrame();
```

**Comment by Eightvo** [comment]
I'm having a bit of an issue implementing this workaround...
so, condensed it's a bit like 
```
                  
                         ImGui.EndFrame();
                            var io = ImGui.GetIO();
                            io.Fonts.Clear();
                            bool defaultReq = true;
                            _prevSystemFont = Config.ActiveConfiguration.GuiStyle.SystemFont;
                            if (!String.IsNullOrEmpty(_prevSystemFont))
                                Config.ActiveConfiguration.GuiStyle.SystemFontPtr = io.Fonts.AddFontFromFileTTF(_prevSystemFont, 13.0f);
                            else
                            {
                                defaultReq = false;
                                Config.ActiveConfiguration.GuiStyle.SystemFontPtr = io.Fonts.AddFontDefault();
                            }

                            _prevUserFont = Config.ActiveConfiguration.GuiStyle.UserFont;
                            if (!String.IsNullOrEmpty(_prevUserFont))
                                Config.ActiveConfiguration.GuiStyle.UserFontPtr = io.Fonts.AddFontFromFileTTF(_prevUserFont, 13.0f);
                            else
                            {
                                if (defaultReq)
                                    Config.ActiveConfiguration.GuiStyle.UserFontPtr = io.Fonts.AddFontDefault();
                                else
                                    Config.ActiveConfiguration.GuiStyle.UserFontPtr = Config.ActiveConfiguration.GuiStyle.SystemFontPtr;
                            }

                        Env.Gui.RecreateFontDeviceTexture(_graphics);

                           ImGui.NewFrame();
```

The issue is that ImGui.NewFrame(); will fail with an exception: System.AccessViolationException
 After loading a font after switching a font after loading a font.

Sorry. I'll try to explain better and try not ramble. 

I would like there to be two fonts available at any given time (a system font and a user font as evident per code sample). I want to be able to update (at least the user) fonts in ImGui at runtime. There is a built in ImGui control which allows you to select a loaded font. In addition, I have created a control that allows a user to select fonts to load. 

Considering these, the exception occurs as follows.
A) ImGui Font Select shows only ProggyClean.ttf, 13px
B) User select desired font in MyControl
C) Code sample provided is triggered with values SystemFont="" and UserFont='path\to\SomeFile.ttf'
D) ImGuiFont Select shows ProggyClean.ttf,13px and Additional loaded font.
E) User select New Font in Font Selector
F) ImGui responds correctly updating font to new font
G) User selects ProggyClean.ttf in Font Selector
H) ImGui respons correcly updating font to default font
I) User selects new desired font in MyControl
J) Code sample provided is triggered with values SystemFont="" and UserFont='new\path\to\file.ttf'
K) Exception is thrown.


Please note, I can perform steps B,C, and D repeatedly assuming I do not proceed to steps E,F,G and H.
Additionally, I have the same issue when the code was modified to maintain up to 5 fonts loaded at a time (default, user, system, previous user, previous system) because I thought it was possible it was referencing the font that didn't get loaded even though the default font was selected again.

I am using Nuget Pkgs:
  Veldrid 4.6.41
  Veldrid.ImGui 5.69.0

**Comment by Eightvo** [comment]
Shoot, sorry... it was simply missing a call to build the fonts atlas.

should be

ImGui.EndFrame();
var io = ImGui.GetIO();
io.Fonts.Clear();
io.Fonts.AddFontFromFileTTF("C:/Windows/Fonts/ArialUni.ttf", 14);
// Load other fonts
io.Fonts.Build();
renderer.RecreateFontDeviceTexture(_gd);
ImGui.NewFrame();

I think.

**Comment by Eightvo** [comment]
I did find a work around that works.

```
ImGui.EndFramer();
GuiRenderer=new ImGuiRenderer(graphics, etc...);
var io = ImGui.GetIO();
io.Fonts.Clear();
use io.Fonts.AddFontFromFileTTF but rember the index instead of the returned pointer.
io.Fonts.Build();
GuiRenderer.RecreateFontDeviceTexture(graphics);
ImGui.NewFramer();
````

Then, when trying to push a font

```
   ImGui.PushFont(ImGui.GetIO().Fonts.Fonts[fontIndx]);
````

**Comment by GrandyB** [comment]
Has this been addressed between time it was raised and now?
Looking to use an icon font merged into the default font, and I have a sinking feeling it might be down to this.

**Comment by zaafar** [comment]
Unfortunately, no.
